---
title: "RNA-seq Differential Expression Analysis Report (`r name`)"
author: "[Genomics Core (Cedars-Sinai)](https://www.cedars-sinai.edu/Research/Research-Cores/Genomics-Core/)"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  html_document: default
  pdf_document: default
params:
  data: "MO-5333--08--14--2018_COUNTS.csv"
  info: "AP-5782--11--08--2018_sample_info.csv"
  comparison: "AP-5782--11--08--2018_comparisons.csv"
  project: "AP-5782--11--08--2018"
  cutoff1: "padj<0.01"
  cutoff2: "1"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an **R Markdown** document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

```{r logo, echo=FALSE}
#htmltools::img(src = knitr::image_uri(file.path("/common/genomics-core/apps/multiqc/Genomics_Core_Logo.jpg")),
#               alt = 'logo',
#               style = 'position:absolute; top:0; right:0; padding:10px;')
#htmltools::img(src = knitr::image_uri(file.path("/home/genomics/genomics/apps/multiqc/Genomics_Core_Logo.jpg")),
#               alt = 'logo',
#               style = 'position:absolute; top:0; right:0; padding:10px;')
```


```{r init, message=FALSE, echo=FALSE}
#Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code.
#library(vegan)
#library(DESeq2)
#library(gplots)
#library(ggplot2)
#library(rgl)
#library(rglwidget)
#library(genefilter)
#library(plot3D)
#library(FactoMineR)
#library(RColorBrewer)
#library(pheatmap)
#library(DT) # Interactive HTML tables
#library(ggpubr)
#library(plotly)
#library(heatmaply)
#library(manhattanly) # Interactive Valcano plot
#setwd("/Users/wud3/Documents/R_analysis/RNAseq/Oct_test/")
```


### - Data Overview
* We start by reading in the data. There are `r dim(data)[1]` genes of `r dim(countdata_1)[2]` samples in the project. Samples are labeled as `r colnames(countdata_1)`, representing `r length(levels(condition))` (i.e. `r paste(levels(condition),collapse=", ")`) groups.
* After removing genes with expression equal to 0, we have `r dim(countdata_1)[1]` genes for further analysis.


### - Principal Component Analysis for RNAseq samples

* PCA plots of RNA-seq data, based on the top 500 genes by variance across all samples were generated. "`r levels(condition)[1]`" and "`r levels(condition)[2]`" groups are represented in red and blue, repsectively.
\
```{r interactive PCA, echo=FALSE, message=FALSE, fig.height = 5, fig.width = 16}
intgroup="condition"
pca_coord <- data.frame(res.pca_1$ind$coord)
intgroup.df <- as.data.frame(colData(rld_1)[, intgroup,drop = FALSE])
percentVar <- res.pca_1$eig[1:3, 2]
lab.x<-paste("PC1 (",format(percentVar[1],nsmall=2,digits=2),"%)",sep = "")
lab.y<-paste("PC2 (",format(percentVar[2],nsmall=2,digits=2),"%)",sep = "")
lab.z<-paste("PC3 (",format(percentVar[3],nsmall=2,digits=2),"%)",sep = "")
PCA1_2 <- plot_ly(pca_coord, x = pca_coord$Dim.1, y = pca_coord$Dim.2, text = rownames(pca_coord), color = intgroup.df$condition, colors = rainbow(length(levels(factor(condition))), s=0.5), marker = list(size = 10), showlegend = FALSE) %>%
  add_markers() %>%
  layout(xaxis = list(title = lab.x, zeroline = FALSE),yaxis = list(title = lab.y, zeroline = FALSE))
PCA1_3 <- plot_ly(pca_coord, x = pca_coord$Dim.1, y = pca_coord$Dim.3, text = rownames(pca_coord), color = intgroup.df$condition, colors = rainbow(length(levels(factor(condition))), s=0.5), marker = list(size = 10), showlegend = FALSE) %>%
  add_markers() %>%
  layout(xaxis = list(title = lab.x, zeroline = FALSE),yaxis = list(title = lab.z, zeroline = FALSE))
PCA2_3 <- plot_ly(pca_coord, x = pca_coord$Dim.2, y = pca_coord$Dim.3, text = rownames(pca_coord), color = intgroup.df$condition, colors = rainbow(length(levels(factor(condition))), s=0.5), marker = list(size = 10)) %>%
  add_markers() %>%
  layout(xaxis = list(title = lab.y, zeroline = FALSE),yaxis = list(title = lab.z, zeroline = FALSE))
p <- subplot(PCA1_2, PCA1_3, PCA2_3, nrows = 1, titleX = TRUE, titleY = TRUE, margin = c(0.05, 0.05, 0.05, 0.05))
#p
p %>% layout(annotations = list(
 list(x = 0.16 , y = 1.05, text = "PC1 vs PC2", showarrow = F, xref='paper', yref='paper'),
 list(x = 0.52 , y = 1.05, text = "PC1 vs PC3", showarrow = F, xref='paper', yref='paper'),
 list(x = 0.89 , y = 1.05, text = "PC2 vs PC3", showarrow = F, xref='paper', yref='paper'))
)
```



* 3D_PCA was generated using the 500 genes that showing the highest variation. `r levels(condition)[1]` and `r levels(condition)[2]` groups are represented in red and blue, repsectively.
\
```{r interactive plot,echo=FALSE, message=FALSE, fig.height = 6, fig.width = 6}
pca <- prcomp(t(assay(rld_1)[select_1, ]),scale. = TRUE)
percentVar <- pca$sdev^2/sum(pca$sdev^2)
intgroup="condition"
intgroup.df <- as.data.frame(colData(rld_1)[, intgroup,drop = FALSE])
d <- data.frame(PC1 = pca$x[, 1], PC2 = pca$x[, 2], PC3 = pca$x[,3],condition=intgroup.df)
lab.x<-paste("PC1 (",format(percentVar[1]*100,nsmall=2,digits=2),"%)",sep = "")
lab.y<-paste("PC2 (",format(percentVar[2]*100,nsmall=2,digits=2),"%)",sep = "")
lab.z<-paste("PC3 (",format(percentVar[3]*100,nsmall=2,digits=2),"%)",sep = "")
p <- plot_ly(d, x = d$PC1, y = d$PC2, z = d$PC3, text = rownames(pca$x), color = d$condition, colors = rainbow(length(levels(factor(d$condition))), s=0.5)) %>%
  add_markers() %>%
  layout(scene = list(xaxis = list(title = lab.x),
                     yaxis = list(title = lab.y),
                     zaxis = list(title = lab.z)))
p
```


```{r DEGs_list, echo=FALSE, message=FALSE}
resdata <- eval(parse(text=paste0("resdata", i)))
names(resdata)[1] <- "Gene"
names(resSig)[1] <- "Gene"
dat.norm_2=as.matrix(norm_1[resSig$Gene,])
gene_list <- gsub(".*_","",resSig$Gene) #remove all and before up to "_"
resSig$Gene <- gene_list
colnames(resSig)[3]<-"log2FC"
colnames(resSig)[5]<-"p_adj"
gene_list <- gsub(".*_","",resdata$Gene) #remove all and before up to "_"
resdata$Gene <- gene_list
colnames(resdata)[3]<-"log2FC"
colnames(resdata)[5]<-"p_adj"
```
### - List of significant DEGs (differentially expressed genes)
* Significance definition
     + `r sig_DEGs_padj`
     + \|FC\|  >= `r sig_DEGs_FC`



* `r dim(resSig)[1]` genes were identified as significant DEGs (`r name`; `r sig_DEGs`), where `r dim(subset(resSig, log2FC>=0))[1]` were upregulated and `r dim(subset(resSig, log2FC<0))[1]` were downregulated.
* The table is showing all genes.  Please check CSV file for all details.
     + basemean = average expression value across all samples
     + p_adj = adjusted p value
\
\
```{r table, echo=FALSE, message=FALSE, warning=FALSE,fig.height = 6, fig.width = 6}
datatable(format(resdata[, 1:5], digits=2, justify="centre"), caption = htmltools::tags$caption(
    style = 'caption-side: top; text-align: center;',
    'Table 1: ', htmltools::em('List of DEGs')))
```

### - Heatmap of DEGs across all samples
* `r dim(dat.norm_2)[1]` genes with significant variances across the `r dim(dat.norm_2)[2]` samples were utilized to make the heatmap.
*  Zoom in to see gene names.
\
\
```{r interactive_heatmap, echo=FALSE, message=FALSE, warning=FALSE, fig.height = 10, fig.width = 10}
#resdata <- merge(as.data.frame(res), as.data.frame(norm), by="row.names", sort=FALSE)
#names(resdata)[1] <- "Gene"
#resSig=subset(resdata, resdata$padj<0.05)
#dat.norm=as.matrix(norm[resSig$Gene,])
#zscore_by_row <- function(x, na.rm = TRUE) {retval <- list()
#         retval$rowMeans <- rm <- rowMeans(x, na.rm = na.rm)
#         x <- sweep(x, 1, rm)
#         retval$rowSDs <- sx <- apply(x, 1, sd, na.rm = na.rm)
#         x <- sweep(x, 1, sx, "/")}
# zscore_data <- zscore_by_row(dat.norm)
#calculate z-score
colv=as.dendrogram(hclust(as.dist(1-cor(dat.norm_2))))
#split by "_" only works for ENSRNOG00000049517_Tnfaip3
#rownames(dat.norm_2)<-matrix(unlist(strsplit(rownames(dat.norm_2),'[_]')),byrow=T,ncol=2)[,2]
#split by the first "_", works for both ENSRNOG00000049517_Tnfaip3 ENSRNOG00000027002_NEWGENE_1310139
rownames(dat.norm_2) <- matrix(unlist(strsplit(sub('(^[^_]+)_(.*)$', '\\1 \\2', rownames(dat.norm_2)), ' ')),byrow=T,ncol=2)[,2]
rowv=as.dendrogram(hclust(as.dist(1-cor(t(dat.norm_2)))))
#p <- heatmaply(dat.norm_2,col=colorRampPalette(c("blue","black","yellow"))(256),k_col=2,Rowv=rowv,dendrogram='row',Colv=colv, scale="row",cexRow =0.5,trace="none",showticklabels=c(TRUE,FALSE),density.info="none",keysize=.8,key.xlab="")

#############################set limit from -2 to 2###############################
temp <- scale(t(dat.norm_2))
temp[temp>2] <- 2
temp[temp< -2] <- -2
p <- heatmaply(t(temp),col=colorRampPalette(c("blue","black","yellow"))(256),k_col=2,labRow = NULL,Rowv=rowv,dendrogram='row',Colv=colv, scale="row",cexRow =0.5,trace="none",density.info="none",keysize=.8,key.xlab="")
#####################################################

#p <- heatmaply(dat.norm_2,col=colorRampPalette(c("blue","black","yellow"))(256),k_col=2,labRow = NULL,Rowv=rowv,dendrogram='row',Colv=colv, scale="row",cexRow =0.5,trace="none",density.info="none",keysize=.8,key.xlab="")
p
# export heatmap as PDF
pdf(file=paste(out_folder,name,"_heatmap.pdf",sep=''),20,25)
#heatmap.2(dat.norm_2,col=colorRampPalette(c("blue","black","yellow"))(256),labRow = FALSE,dendrogram='row',Rowv=rowv,Colv=colnames(dat.norm_2),scale="row",cexRow =0.5,trace="none",density.info="none",keysize=.8,key.xlab="",margins=c(15,14))

######################### set limit from -2 to 2 ##########################
colors = c(seq(-2,-0.11,by=0.03),seq(-0.08,0.13,by=0.03),seq(0.16,2,by=0.03))
my_palette <- colorRampPalette(c("blue", "black", "yellow"))(length(colors)-1)
heatmap.2(dat.norm_2,col=my_palette,labRow = FALSE,dendrogram='row',Rowv=rowv,Colv=colnames(dat.norm_2),scale="row",cexRow =0.5,trace="none",density.info="none",keysize=.8,key.xlab="",margins=c(15,14),breaks=colors)

dev.off()
```


### - Volcano Plot
* Volcano plot reporting adjusted p-value (−log10(p), y axis) as a function of log2 (fold change) between the groups (logFC, x axis).
* Fold change(`r name`) of the expression level for a given gene was defined as the read counts of the gene in the `r comps[i,][2]` sample divided by the read counts in the `r comps[i,][3]` sample.
* Genes that are identified as significant DEGs with `r sig_DEGs` are colored in red.
\
\
\
```{r interactive Valcano Plot, echo=FALSE, message=FALSE, warning=FALSE,fig.height = 6, fig.width = 6}
# create a volcano object
resdata_1 = subset(resdata, p_adj!="NA")
resdata_1$negative_Log10_padj <- -log10(resdata_1$p_adj)

  #cutoff1: "padj<0.01"
  #cutoff2: "1"

if (sig_DEGs_FC == 1) {
  cutoff1_temp <- gsub(" ", "", sig_DEGs_padj, fixed = TRUE) #delete space from string
  temp <- unlist(strsplit(cutoff1_temp, "<")) #temp[1] is either pvalue or padj; temp[2] is the cutoff
  resdata_1$Colour[resdata_1[,temp[1]] < as.numeric(temp[2])] = paste0("Significant(", cutoff1_temp, ")")
  resdata_1$Colour[resdata_1[,temp[1]] >= as.numeric(temp[2])] = "Non-significant"
} else if (sig_DEGs_FC > 1) {
  cutoff1_temp <- gsub(" ", "", sig_DEGs_padj, fixed = TRUE) #delete space from string
  temp <- unlist(strsplit(cutoff1_temp, "<")) #temp[1] is either pvalue or padj; temp[2] is the cutoff
  resdata_1$Colour[(abs(resdata_1$log2FC) >= log2(as.numeric(sig_DEGs_FC))) & (resdata_1[,temp[1]] < as.numeric(temp[2]))]= paste0("Significant(", cutoff1_temp, "|FC|>=", sig_DEGs_FC,")")
  #"Significant(padj<0.01,|FC|>=2)"
  resdata_1$Colour[(abs(resdata_1$log2FC) < log2(as.numeric(sig_DEGs_FC))) | (resdata_1[,temp[1]] >= as.numeric(temp[2]))]="Non-significant"
  #resdata_1$Colour[(abs(resdata_1$log2FC) >= 1) & (resdata_1$p_adj < 0.01)]="Significant(padj<0.01,|FC|>=2)"
  #resdata_1$Colour[(abs(resdata_1$log2FC) < 1) | (resdata_1$p_adj >= 0.01)]="Non-significant"
} else {
  print("cutoff value is not appropreate")
}
#print("cutoff value is not appropreate")


p <- plot_ly(resdata_1, x = ~log2FC, y = ~negative_Log10_padj, type = 'scatter',
        mode = "markers", color = ~Colour, colors = c("black", "red"), text = ~paste('Gene: ', Gene))%>%
  layout(title = name, #name <- paste0(comps[i,][2], "_vs_", comps[i,][3])
         yaxis = list(zeroline = FALSE),
         xaxis = list(zeroline = FALSE))
p
library(ggrepel)
pdf(file=paste0(out_folder,name,"_Volcano.pdf"),14,10)
resdata_1$label<-""  #####this is to save the label of your interested genes
p<-ggplot(resdata_1,aes(x=log2FC,y=negative_Log10_padj,label=label))
p<-p+geom_point(aes(color = resdata_1$Colour), size = 0.8)+scale_color_manual(values = c("black", "red"))
p<-p+theme(panel.background = element_rect(fill = "white", colour = "black"))+geom_label_repel(point.padding = unit(0.1, "lines"))+labs(x = paste0("log2FoldChange (", name, ")"), y = "-log10(adjusted p-value)", title = name, color = "")
p
dev.off()
```


### - MA plot
* MA plot for differential expression analysis (where M = log ratios and A = mean values).
* For each gene, the log2(fold change) (log2(`r paste0(comps[i,][2], "/", comps[i,][3])`)) between the two groups is plotted (A, y axis) against the gene's log2(average expression) in the two groups (M, x axis).
* Genes that are identified as significant DEGs with `r sig_DEGs` are colored in red (significant upregulated genes) and in blue (significant downregulated genes).
\
```{r interactive MA_plot, echo=FALSE, message=FALSE, fig.height = 6, fig.width = 6, warning=FALSE}
#volcanorObj <- volcanor(resdata1, )
#volcanoly(resdata1, p=resdata1$padj)
library(ggplot2)
library(ggthemes)


if (sig_DEGs_FC == 1) {
  cutoff1_temp <- gsub(" ", "", sig_DEGs_padj, fixed = TRUE) #delete space from string
  temp <- unlist(strsplit(cutoff1_temp, "<")) #temp[1] is either p-value or padj; temp[2] is the cutoff
  resdata_1$Colour[(resdata_1$log2FC > 0) & (resdata_1[,temp[1]] < as.numeric(temp[2]))] = paste0("Significant UpR(", cutoff1_temp, ")")
  resdata_1$Colour[(resdata_1$log2FC < 0) & (resdata_1[,temp[1]] < as.numeric(temp[2]))] = paste0("Significant DownR(", cutoff1_temp, ")")
  resdata_1$Colour[resdata_1[,temp[1]] >= as.numeric(temp[2])] = "Non-significant"
} else if (sig_DEGs_FC > 1) {
  cutoff1_temp <- gsub(" ", "", sig_DEGs_padj, fixed = TRUE) #delete space from string
  temp <- unlist(strsplit(cutoff1_temp, "<")) #temp[1] is either p-value or padj; temp[2] is the cutoff
  resdata_1$Colour[(resdata_1$log2FC >= log2(as.numeric(sig_DEGs_FC))) & (resdata_1[,temp[1]] < as.numeric(temp[2]))]= paste0("Significant UpR(", cutoff1_temp, "FC>=", sig_DEGs_FC,")")
  resdata_1$Colour[(resdata_1$log2FC <= -log2(as.numeric(sig_DEGs_FC))) & (resdata_1[,temp[1]] < as.numeric(temp[2]))]= paste0("Significant DownR(", cutoff1_temp, "FC<=-", sig_DEGs_FC,")")
} else {
  print("cutoff value is not appropreate")
}
#print("cutoff value is not appropreate")



resdata_1$log2Mean <- log2(resdata_1$baseMean + 1)
plot <- ggplot(data=resdata_1, aes(x=log2Mean, y=log2FC, colour=Colour)) + geom_point(aes(text = paste("Gene:", resdata_1$Gene)), size = 1) + scale_color_manual( values=c("dark gray", "blue", "red")) + labs(x = "Log2 Mean_Expression", y = "Log2 Fold_Change", title = name)
gg <- ggplotly(plot)
gg
#Export MA plot as a PDF file
pdf(file=paste0(out_folder,name,"_MAplot.pdf"),16,9)
resdata_1$label<-"" #####this is to save the label of your interested genes
p<-ggplot(resdata_1,aes(x=log2Mean, y=log2FC,label=label))
p<-p+geom_point(aes(color = resdata_1$Colour), size = 0.8)+scale_color_manual(values = c( "grey","blue","red"))
p<-p+geom_hline(yintercept = c(0, -1, 1), linetype = c(1, 2, 2), color = c("black", "black", "black"))
p<-p+theme(panel.background = element_rect(fill = "white", colour = "black"))+geom_label_repel(force = 1,box.padding = unit(0.35, "lines"),point.padding = unit(0.3,"lines"))
p<-p+labs(x = "Log2 mean expression", title = name, color = "")
p
dev.off()
```
